federation/pkg/federation-controller/replicaset/replicasetcontroller.go

```diff
@@ -0,0 +1,461 @@
+		frsc.replicasetWorkQueue.Add(item.Key)
+	})
+	frsc.clusterDeliverer.StartWithHandler(func(_ *fedutil.DelayingDelivererItem) {
+		frsc.reconcileNamespacesOnClusterChange()
+	})
+
+	for i := 0; i < workers; i++ {
+		go wait.Until(frsc.worker, time.Second, stopCh)
+	}
+	<-stopCh
+	glog.Infof("Shutting down ReplicaSetController")
+	frsc.replicasetDeliverer.Stop()
+	frsc.clusterDeliverer.Stop()
+	frsc.replicasetWorkQueue.ShutDown()
+
+}
+
+func (frsc *ReplicaSetController) isSynced() bool {
+	if !frsc.fedInformer.ClustersSynced() {
+		glog.V(2).Infof("Cluster list not synced")
+		return false
+	}
+	clusters, err := frsc.fedInformer.GetReadyClusters()
+	if err != nil {
+		glog.Errorf("Failed to get ready clusters: %v", err)
+		return false
+	}
+	if !frsc.fedInformer.GetTargetStore().ClustersSynced(clusters) {
+		return false
+	}
+	if !frsc.replicaSetController.HasSynced() {
+		glog.V(2).Infof("federation replicaset list not synced")
+		return false
+	}
+	return true
+}
+
+func (frsc *ReplicaSetController) deliverLocalReplicaSet(obj interface{}, duration time.Duration) {
+	key, err := controller.KeyFunc(obj)
+	if err != nil {
+		glog.Errorf("Couldn't get key for object %v: %v", obj, err)
+		return
+	}
+	_, exists, err := frsc.replicaSetStore.GetByKey(key)
+	if err != nil {
+		glog.Errorf("Couldn't get federation replicaset %v: %v", key, err)
+		return
+	}
+	if exists { // ignore replicasets exists only in local k8s
+		frsc.deliverReplicaSetByKey(key, duration, false)
+	}
+}
+
+func (frsc *ReplicaSetController) deliverFedReplicaSetObj(obj interface{}, delay time.Duration) {
+	key, err := controller.KeyFunc(obj)
+	if err != nil {
+		glog.Errorf("Couldn't get key for object %+v: %v", obj, err)
+		return
+	}
+	frsc.deliverReplicaSetByKey(key, delay, false)
+}
+
+func (frsc *ReplicaSetController) deliverReplicaSetByKey(key string, delay time.Duration, failed bool) {
+
+	if failed {
+		frsc.replicaSetBackoff.Next(key, time.Now())
+		delay = delay + frsc.replicaSetBackoff.Get(key)
+	} else {
+		frsc.replicaSetBackoff.Reset(key)
+	}
+	frsc.replicasetDeliverer.DeliverAfter(key, nil, delay)
+}
+
+func (frsc *ReplicaSetController) worker() {
+	for {
+		item, quit := frsc.replicasetWorkQueue.Get()
+		if quit {
+			return
+		}
+		key := item.(string)
+		err := frsc.reconcileReplicaSet(key)
+		frsc.replicasetWorkQueue.Done(item)
+		if err != nil {
+			glog.Errorf("Error syncing cluster controller: %v", err)
+			frsc.deliverReplicaSetByKey(key, 0, true)
+		}
+	}
+}
+
+func (frsc *ReplicaSetController) schedule(frs *extensionsv1.ReplicaSet, clusters []*fedv1.Cluster,
+	expected map[string]int64, actual map[string]int64) map[string]int64 {
+	// TODO: integrate real scheduler
+	replicas := *frs.Spec.Replicas
+	var clusterNames []string
+	for _, cluster := range clusters {
+		clusterNames = append(clusterNames, cluster.Name)
+	}
+	scheduleResult := frsc.planner.Plan(int64(replicas), clusterNames)
```

Comment:
Do you mean build a planner for every reconciliation?
I thought the preference passed to the planner is for default behavior. 
We are having a separate scheduler PR that parse the annotation and calls planner (probably modified) and this switch to that once its ready.